packages/nextjs/src/utils/handlers.ts

```diff
@@ -24,73 +30,113 @@ export const withSentry = (handler: NextApiHandler): WrappedNextApiHandler => {
-
-            const transaction = startTransaction(
-              {
-                name: `${reqMethod}${reqPath}`,
-                op: 'http.server',
-                ...traceparentData,
-              },
-              // extra context passed to the `tracesSampler`
-              { request: req },
-            );
-            currentScope.setSpan(transaction);
           }
+          const reqMethod = `${(req.method || 'GET').toUpperCase()} `;
+
+          const transaction = startTransaction(
+            {
+              name: `${reqMethod}${reqPath}`,
+              op: 'http.server',
+              ...traceparentData,
+            },
+            // extra context passed to the `tracesSampler`
+            { request: req },
+          );
+          currentScope.setSpan(transaction);
+
+          // save a link to the scope on the response, so that even if there's an error (landing us outside of
+          // the domain), we can still finish the transaction and attach the correct data to it
+          (res as ScopedResponse).__sentryScope = currentScope;
         }
+      }
 
+      try {
         return await handler(req, res); // Call original handler
       } catch (e) {
-        withScope(scope => {
-          scope.addEventProcessor(event => {
+        if (currentScope) {
+          currentScope.addEventProcessor(event => {
             addExceptionMechanism(event, {
               handled: false,
             });
-            return parseRequest(event, req);
+            return event;
           });
           captureException(e);
-        });
-        throw e;
-      } finally {
-        const transaction = getActiveTransaction();
-        if (transaction) {
-          transaction.setHttpStatus(res.statusCode);
-
-          transaction.finish();
-        }
-        try {
-          await flush(2000);
-        } catch (e) {
-          // no-empty
         }
+        throw e;
       }
     });
 
     return await boundHandler();
   };
 };
+
+type ResponseEndMethod = ScopedResponse['end'];
+type WrappedResponseEndMethod = ResponseEndMethod;
+
+function wrapEndMethod(origEnd: ResponseEndMethod): WrappedResponseEndMethod {
+  return async function newEnd(this: ScopedResponse, ...args: unknown[]) {
+    // if the handler errored, it will have popped us out of the domain, so push the domain's scope onto the stack
+    // just in case (if we *are* still in the domain, this will replace the current scope with a clone of itself,
+    // which is effectively a no-op as long as we remember to pop it off when we're done)
+    const currentHub = getCurrentHub();
+    currentHub.pushScope(this.__sentryScope);
+
+    const transaction = getActiveTransaction();
+
+    if (transaction) {
+      transaction.setHttpStatus(this.statusCode);
+
+      // Push `transaction.finish` to the next event loop so open spans have a better chance of finishing before the
+      // transaction closes, and make sure to wait until that's done before flushing events
```

Comment:
What's an example of "open span" here?

Reply:
Hmm, I guess any async calls wrapped with spans.
Hard to imagine every user having to deal with this kind of trick for their own instrumentation.

Reply:
See https://github.com/getsentry/sentry-docs/pull/3862, for example. This is something we do in [our Express handlers already](https://github.com/getsentry/sentry-javascript/blob/f8a4d60e88372b1e8592a24c0ebc59f30e9f4c6a/packages/node/src/handlers.ts#L83-L85) as well.